### ISSUE TITLE
Use `name.namespace.html` as the canonical URL, not `kind.name.html`

### ISSUE THREAD

**Issue by scottmcm** [rust-lang/rust#55160]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/issues/55160)*

I just wanted to look at the `ManuallyDrop` docs again, and my browser helpfully remembered that it's at <https://doc.rust-lang.org/std/mem/union.ManuallyDrop.html>.  But, despite the type being stable, that page no longer exists; it's now <https://doc.rust-lang.org/std/mem/struct.ManuallyDrop.html>.

Is the specific item kind actually important to have in the path?  The namespace (type vs value vs ...) certainly is, but especially for things without public fields, making the struct-vs-union distinction so front-and-centre seems overall unhelpful.

**Comment by GuillaumeGomez** [comment]
For me it's a positive thing. I'd argue that you'd better not use direct URLs and just search or rustdoc links. The current URLs are clear and represent what they should.

**Comment by mikeyhew** [comment]
@GuillaumeGomez I disagree with that, I think that direct links to documentation are useful, their use should be encouraged, and we should avoid breaking them if possible.

Having the extra information of `struct`, `enum` or `type`  in the URL is kind of nice too, but the other thing is that switching from the current URLs to new ones would result in a lot of links breaking, so maybe there is some way to use redirects instead? Like if you went to https://doc.rust-lang.org/std/mem/union.ManuallyDrop.html or https://doc.rust-lang.org/std/mem/type.ManuallyDrop.html it would redirect you to https://doc.rust-lang.org/std/mem/struct.ManuallyDrop.html.

**Comment by GuillaumeGomez** [comment]
I'm not sure we understood each other. To be clear: the current state of rustdoc URLs seems good and should be kept unless something way better is proposed.

**Comment by RalfJung** [comment]
I agree with the OP that encoding the "kind" of type in the URL seems suboptimal. Rust has three namespaces (macro-level, fn/variable-level, type-level), so that needs to be encoded, but I see no good reason to encode more. `type.ManuallyDrophtml` has all the information needed to uniquely identify this type.

As another argument for why the current URLs are bad: this is one of the few cases where doing a semver-compatible change (changing a `struct` to an `enum`, for example) breaks docs links.

**Comment by nox** [comment]
Changing a struct to an enum isn't backwards-compatible (one can match against `Struct { .. }`), but I support removing the type kind from URLs, for the same rationale as in #64454.

**Comment by bluss** [comment]
We have a bunch of redirects that only focus on the namespace.

For example the URL https://doc.rust-lang.org/stable/std/mem/ManuallyDrop.t.html with "t" for type, will redirect to the correct place and I assume this URL was there before the struct/union switch happened. Only problem is that it's not the canonical URL.  (And, it would be more ideal if we didn't have lots of extra redirect files, there are enough small files already in a rustdoc site).

**Comment by RalfJung** [comment]
@bluss wow, I had no idea these files even existed!

For use in rustc/libstd, if they should be useful we need a tidy pass that tells people to link to the "t" file instead of the "struct"/"enum"/"union" file.

**Comment by GuillaumeGomez** [comment]
It was for RLS iirc. Not sure if removing it is fine but I'd like to remove it.

**Comment by jyn514** [comment]
> I disagree with that, I think that direct links to documentation are useful, their use should be encouraged, and we should avoid breaking them if possible.

I have the same comment as in https://github.com/rust-lang/rust/issues/71912#issuecomment-680097427: direct links should use intra-doc links, the links rustdoc generates are not stable.

**Comment by RalfJung** [comment]
@jyn514 so say I am writing a blog post and mentioning some Rust types whose docs I want to reference. Obviously my site generator does not support intra-doc links. What do you propose I do?

This is something people already do -- anything that has a URL has people link to it. It's what makes the web so amazing. Please don't break that.

**Comment by GuillaumeGomez** [comment]
URLs are stable, the DOM content isn't. However, since you always use a specific version of a crate, you can just link to the element of this specific version of the crate on docs.rs.

**Comment by RalfJung** [comment]
URLs are not stable when a `struct` becomes an `enum` (which otherwise is semver-compatible if the struct had a private field).

**Comment by GuillaumeGomez** [comment]
Then that's on the crate owner, not on rustdoc. We the URLs are defined as `[type].[type name].html`. If you change an item's type, then you change the URL. It's predictable. But again: always reference to a specific version of a crate and that shouldn't be an issue.

**Comment by RalfJung** [comment]
The entire point of this issue is that there is no good reason to include struct vs union in the URL. The "item type" of all structs, unions and enums should be just `type`.

It's predictable, sure, but that does not make it sensible.

**Comment by GuillaumeGomez** [comment]
So you would want to include a breaking change for something stable even before 1.0? You'd need very strong reasons for it to be done and I'm really not convinced...

**Comment by RalfJung** [comment]
No we don't need a breaking change -- as @bluss wrote above, reasonably stable URLs already exist: https://doc.rust-lang.org/stable/std/mem/ManuallyDrop.t.html
But nobody knows about them because they are not used by default. So all it would take is swap the role of `ManuallyDrop.t.html` and `struct.ManuallyDrop.html`, making the latter a redirect and the former the target of all the automatically generated links.

**Comment by GuillaumeGomez** [comment]
Well, we kinda want to remove this URL format since a while. It's still there because it's being used by RLS iirc. Also, this would be a regression in my point of view considering you lose information. And since it's not the default URL format, it's definitely a breaking change.

**Comment by jyn514** [comment]
> So you would want to include a breaking change for something stable

The point of discussion IMO is whether or not is _is_ stable. Currently, it's stable with respect to _rustdoc_ versions, but not with respect to the code being documented. Either

a) The file names should be predictable across versions being documented, in which case we should recommend `ManuallyDrop.t.html`, or
b) The file names are unstable and no one should depend on them, in which case we should change them every release so people don't depend on them by accident.

The current middle ground is sort of the worst of both worlds because it _looks_ like it's stable, but can break for changes that do not break semver.

**Comment by RalfJung** [comment]
The information you lose is not semver-stable and irrelevant for API consumers. That's what this issue is all about -- please consult the arguments given above.

**Comment by RalfJung** [comment]
IMO the question is quite simple at a fundamental level -- what does it take to uniquely identify an item in Rust? You needs its name, its full module path, and its "kind" -- which can be value, type, or macro. And that's it. So this information should be sufficient to construct a URL. After all, a "Uniform Resource Locator" is all about being able to locate resources, such as the documentation about items in Rust, in a uniform and consistent way.

**Comment by jyn514** [comment]
@RalfJung what do you propose we do about all the existing URLs that point to `enum.ManuallyUninit.html`? Should we generate redirects for _all_ pages in that namespace? That seems like it would bloat the size of the docs quite a bit. But if we don't do that, then the URLs are still broken even if we make `union.ManuallyUninit.html` a redirect.

**Comment by RalfJung** [comment]
If my understanding is correct, currently we generate two pages for every type:
* `[enum|struct|union].$name.html` is the main page with all the fields and methods and stuff
* `$name.t.html` redirects to the above.

What I am proposing is to swap the role of these two files, so every existing link keeps working:
* `$name.t.html` is the main page with all the fields and methods and stuff
*  `[enum|struct|union].$name.html` redirects to the above.

This cannot break any link that currently works. It means changing e.g. union to enum still breaks people that use the "wrong" (deprecated) `union.$name.html`, but that's certainly better than not having the `union.$name.html` redirect! Currently we break *all* links on a union -> enum change; with the new scheme we only break old links -- but new links that people set will go to the `.t.html` page. So, I'd say this is a strict improvement over the status quo.

**Comment by jyn514** [comment]
I'm happy with that. It doesn't fix the existing links to `union.ManuallyInit.html`, but it encourages good practice for future links without being too much of a burden on rustdoc.

**Comment by RalfJung** [comment]
(I am not sure what `ManuallyInit` is. `ManuallyDrop` changed from union to struct, so if you mean the existing links to `union.ManuallyDrop.html` then I agree.)

**Comment by jyn514** [comment]
I was actually thinking of `MaybeUninit` 😆 so not even close

**Comment by RalfJung** [comment]
`MaybeUninit` is a union and always was, so I don't think it is relevant here.^^

**Comment by ollie27** [comment]
The namespace based redirects were removed in #70563.

**Comment by jyn514** [comment]
It looks like that was mostly to fix wonky hash handling? The only comment about namespaces I see is https://github.com/rust-lang/rust/pull/70563#issuecomment-652653459:

> We should also fully revert #35236 not just part of it.

Is there any reason _not_ to add the namespace redirects back?

**Comment by GuillaumeGomez** [comment]
Why would we want it? It's a duplication of the current system. We removed it to stop having to maintain it and therefore reduce the amount of code.

**Comment by jyn514** [comment]
The _entire issue_ is giving reasons why we would want it. https://github.com/rust-lang/rust/issues/55160#issuecomment-681083071 has a good summary.

**Comment by RalfJung** [comment]
> Why would we want it?

Please read this issue.

**Comment by Kixunil** [comment]
@nox turns out [you're right](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=db88d3447131aabd4ed5a2b618510200), I'm glad I double-checked it. I think this is a mistake in the language design and should be deprecated. It's very impractical and I was unknowingly relying on it not being the case.

The comment I originally intended to write, still relevant **if the incompatibility gets fixed**:

Somewhat less related but IMO **if the type has no public fields** people shouldn't even know what "kind" the type is. Whether it's a tuple struct or ordinary struct or an enum with private variants, if Rust ever gets those, should be completely irrelevant to the consumer because it has no observable effects. This information shouldn't be present in URL or docs or anywhere outside the source code. It's exactly like the analogy with having line numbers in the docs.

**Comment by RalfJung** [comment]
@rust-lang/rustdoc this has caused trouble again recently when `NonZeroU64` etc types where turned into type aliases -- now all the old links to `struct.NonZeroU64.html` are broken. I think every case where a semver-compatible update breaks links to docs is a rustdoc bug.

@GuillaumeGomez 4 years ago you asked for reasons about why we'd want to use different URLs. Examples have been given in this issue both before and after you raised that point. So now that we have answered your question, would be great to hear your thoughts on this issue. :)

**Comment by Kixunil** [comment]
That's what I meant by "if Rust ever gets those". :)

**Comment by lolbinarycat** [comment]
> There are no enums with private variants. However if we ever get such a feature, then I agree.

Rust actually *used* to have enums with private variants, before 1.0, however they were removed because they were the only thing keeping the `priv` keyword in the language.

**Comment by hanna-kruppe** [comment]
What’s the problem with case insensitive file systems?

**Comment by RalfJung** [comment]
> The only remaining issue is for users relying on URLs, but for them, I don't really see a good solution.

We have a great solution: just use a URL that only contains information that semver mandates to be stable.

> It increases the likelihood of having duplicates (because of less namespaces). 

So you mean e.g. if there is an enum and a struct whose name differs only by the case? Do you have any reason to believe that this is more likely than having two structs whose name differs only by case? "Put the kind of ADT into the URL" is not a solution to this problem. It just helps sometimes, while also making rustdoc URLs unsuited for bookmarking or any sort of other reference from the rest of the web.

Between "having stable URLs one can actually rely on" and "sometimes successfully paper over case insensitivity without actually solving the problem", why would you prefer the latter?

**Comment by RalfJung** [comment]
> It means not breaking all existing URLs (on docs.rs for example)

This has been discussed above -- no breakage is needed, the existing URLs can keep working as redirects.

> It works all the time as long as you use the crate version in the URL (in docs.rs)

> It's easy to go around this limitation by using ?search=WhateverYouNeed

The default way to grab a URL is to just copy it from your browser. Both for the std docs and for docsrs, that will not give a stable URL. Only a tiny fraction of people will know to even do anything by hand here, and even ewer will bother doing so if it requires manual work.

In my estimation, ~everyone linking to std docs uses the `stable` or `nightly` URLs. There's really no reason not to, unless you happen to know about this rustdoc issue. "In theory people could avoid the problem" is really not helpful advice when in practice, nobody knows they have to avoid it.

**Comment by hanna-kruppe** [comment]
As described earlier, the URLs for the *stable* `NonZero{U,I}{8,16,32,64,size}` types recently changed because those types were distinct structs before but are now type aliases (`type NonZeroT = NonZero<T>;`).

**Comment by GuillaumeGomez** [comment]
😮 

I wonder if changing the type of an item is considered a semver breaking change...

**Comment by hanna-kruppe** [comment]
It’s not, at least not in this case, as evidenced by the fact that Rust did it.

**Comment by its-the-shrimp** [comment]
Changing from an enum should be a breaking change, since one can always match on its variant, there are no more private variants. However changing from any type kind to a type alias is always backwards-compatible, as is going from a struct with all fields private to an enum

**Comment by lolbinarycat** [comment]
> as is going from a struct with all fields private to an enum

As mentioned before, this is untrue, since you can match against `Struct { .. }`.

**Comment by GuillaumeGomez** [comment]
Ok, then time for an RFC I guess.

**Comment by jyn514** [comment]
cc https://github.com/rust-lang/rfcs/issues/2988

**Comment by hanna-kruppe** [comment]
This issue still being unsolved means the upcoming stabilization of the never type will break links like https://doc.rust-lang.org/std/convert/enum.Infallible.html because part of the stabilization is changing `type Infallible = !;`

@GuillaumeGomez you said “time for an RFC” but who’s going to write that RFC? What questions need to be answered by that RFC? Is it just saying  “rustdoc URLs should be based on the kind of namespace rather than kind of item, and we’ll redirect the old URLs” or is there non-obvious design choices to be made?

Edit: @jyn514’s old RFC is mostly straightforward, I guess at minimum someone needs to pick up that RFC and incorporate the remaining feedback. But I wonder if there’s any other concerns/complexities since then.

**Comment by GuillaumeGomez** [comment]
Adding it to next rustdoc team meeting agenda.

**Comment by cuviper** [comment]
Note that tuple structs also occupy the value namespace for their constructor. I think it's fine if the primary docs are still for the type namespace, but maybe a redirect from the value namespace would be appropriate.

**Comment by camelid** [comment]
> What questions need to be answered by that RFC? Is it just saying “rustdoc URLs should be based on the kind of namespace rather than kind of item, and we’ll redirect the old URLs” or is there non-obvious design choices to be made?

I think one problem is that we still don't have a solution for the filename conflicts on case-insensitive file systems, which is why we have `https://doc.rust-lang.org/nightly/std/keyword.SelfTy.html` instead of `https://doc.rust-lang.org/nightly/std/keyword.Self.html`. This will only get worse if we switch to `name.namespace.html`. However, it's probably still better to switch to the namespace URL style and figure out the case sensitivity issues as an orthogonal issue.
